### PR TITLE
[FEATURE] Client side default values

### PR DESCRIPTION
**Pull request by nyalldawson**
Implements client side default value support for field values. This allows an expression to be set for a vector layer field which is used to evaluate a default value for this field.

Default value expressions can utilise properties of the feature which exist at the time of calling, such as digitized geometries (but not other field values! This is made clear to users by not showing fields and values in the associated expression builders). Expression variables from global, project and layer contexts can be used in default value expressions, making it easier to eg insert a user's name, current datetime, project path, etc

### CODE REVIEW

**Comment by nyalldawson** [comment]
@m-kuhn 

This should be all ready to go - however, there's a related cleanup I'd like to make before merging this.

I think we should split the notion of provider default values into two separate functions, one which returns only constant literal default values and one which returns the default clauses which may be provider specific expressions. These are very different concepts and need to be handled in a different way. What do you think? If you're in favour, I'll do this and make sure QgsVectorLayer::defaultValue only falls back to the literal QgsVectorDataProvider default values.

**Comment by m-kuhn** [comment]
Yay! Thanks. 

I don't understand the plans for the provider default values exactly. Is it about a) the SQL and b) an evaluated literal?
The option "evaluate default values" would be superseded by this new method, correct?

File: python/core/qgsvectorlayer.sip [hunk 1]
```diff
@@ -1202,6 +1202,42 @@ class QgsVectorLayer : QgsMapLayer
...
+     * then properties from the feature (such as geometry) can be used when calculating
+     * the default value.
+     * @param context optional expression context to evaluate expressions again. If not
+     * specified, a default context will be created
+     * @return calculated default value
+     * @note added in QGIS 3.0
+     * @see setDefaultValueExpression()
+     */
+    QVariant defaultValue( int index, const QgsFeature& feature = QgsFeature(),
+                           QgsExpressionContext* context = nullptr ) const;
```
Comment (m-kuhn) [inline python/core/qgsvectorlayer.sip]: So you prefer the full context over just scope(s)?

File: python/core/qgsvectorlayer.sip [hunk 2]
```diff
@@ -1202,6 +1202,42 @@ class QgsVectorLayer : QgsMapLayer
...
+
+    /** Sets an expression to use when calculating the default value for a field.
+     * @param index field index
+     * @param expression expression to evaluate when calculating default values for field. Pass
+     * an empty expression to clear the default.
+     * @note added in QGIS 3.0
+     * @see defaultValue()
+     * @see defaultValueExpression()
+     */
+    void setDefaultValueExpression( int index, const QString& expression );
```
Comment (m-kuhn) [inline python/core/qgsvectorlayer.sip]: Can you add a Q_PROPERTY?

File: src/app/qgisapp.cpp [hunk 3]
```diff
@@ -7343,6 +7343,11 @@ void QgisApp::editPaste( QgsMapLayer *destinationLayer )
     remap.insert( idx, dst );
   }
 
+  QgsExpressionContext context;
+  context << QgsExpressionContextUtils::globalScope()
+  << QgsExpressionContextUtils::projectScope()
+  << QgsExpressionContextUtils::layerScope( pasteVectorLayer );
```
Comment (m-kuhn) [inline src/app/qgisapp.cpp]: `QgsVectorLayer::createContext()` does the same ;)

**Comment by 3nids** [comment]
Wouldn't it make sense to offer by default the edit widget to define the defaut value?
Especially for value relation, relation reference, etc.
We could offer the edit widget by default, with an "advanced" switch to expression widget.
What do you think?

**Comment by nyalldawson** [comment]
> Wouldn't it make sense to offer by default the edit widget to define the defaut value?

I think default values should remain edit-widget agnostic. I honestly can't see someone needing to change them so often that there's a worthwhile benefit from the extra code complexity.

IMO this + constraints should be moved out of the edit widget screen as soon as the main fields panel is redesigned and there's enough room to have these settings there. Then they'll be more closely associated with the fields themselves rather then the edit widget config.

File: python/core/qgsvectorlayer.sip (hunk 1 above)
Comment (nyalldawson) [inline python/core/qgsvectorlayer.sip]: I think so. It addresses a similar use case with a fraction more flexibility (if someone wanted to omit the project / global / layer scopes they could).

**Comment by nyalldawson** [comment]
> I don't understand the plans for the provider default values exactly. Is it about a) the SQL and b) an evaluated literal?

It's about separating these two. Here's my reasoning:

- most of the current use of QgsVectorDataProvider::defaultValue comes from trying to handle when the defaultValue is a different field type from the field - ie a provider side sql string for a numeric field, and making sure it doesn't get formatted as a number. The intention of this code would be easier to understand if it was instead calling a QgsVectorDataProvider::defaultValueClause function
- I hit issues when implementing the default value handling for spatialite since those values are always literals. This causes a side effect where the string literals are either shown in a different appearance in the attribute dialog because they can't be distinguished from provider side default clauses, or shown escaped and surrounded by 's (confusing for users). 

I just think it'll be cleaner to separate these two concepts.

**Comment by 3nids** [comment]
@nyalldawson Would it be possible to display `QgsEditorWidgetFactory::representValue`? Even if they are not changed really often, it is quite useful to see the actual meaning of the value.

**Comment by m-kuhn** [comment]
> I just think it'll be cleaner to separate these two concepts.

That would be the following? Please do

    QVariant defaultValue() const;

    QString defaultValueClause() const;

Do you also have an idea how to avoid avoid unintentional change of default values? i.e. how to avoid a sequence being incremented when its value is not actually used? At the moment I think `defaultValue()` called various times every time a form is opened (both when adding and editing a feature).

File: python/core/qgsvectorlayer.sip (hunk 2 above)
Comment (nyalldawson) [inline python/core/qgsvectorlayer.sip]: Unless I'm mistaken I can't here - the index parameter prevents it.

**Comment by nyalldawson** [comment]
> @nyalldawson Would it be possible to display QgsEditorWidgetFactory::representValue? Even if they are not changed really often, it is quite useful to see the actual meaning of the value.

@3nids ok, i've pushed a change which adds a "preview" label below the default editor which uses representValue to display the results. Note that won't dynamically respond to changes to the widget configuration, but I guess it still has some value if you are making multiple changes to the default expression. It also shows any evaluation/parser error ,so that's also of use.

Also note that I haven't extensively tested this as widget configuration seems broken on master - changes are lost when applying the vector layer properties. ;)

**Comment by nyalldawson** [comment]
> That would be the following? Please do
QVariant defaultValue() const;
QString defaultValueClause() const;
Do you also have an idea how to avoid avoid unintentional change of default values? i.e. how to avoid a sequence being incremented when its value is not actually used? At the moment I think defaultValue() called various times every time a form is opened (both when adding and editing a feature).

@m-kuhn will do - I'll address this in a separate PR

**Comment by nirvn** [comment]
@nyalldawson , it d be great for those default expression / values to be remembered in .qml files so it can be restored when loading a dataset in a new project when such .qml file is present.

**Comment by 3nids** [comment]
@nyalldawson cool, thanks a lot. Did not notice it was broken...

**Comment by nyalldawson** [comment]
@3nids @m-kuhn I've made the default value preview better by adding a changed() signal to QgsEditorConfigWidget, so that the preview is now updated whenever the config changes. Ie, changes are reflected immediately when editing value maps, etc. 

@m-kuhn have you had a chance to look this over yet? Am I ok to merge?

**Comment by m-kuhn** [comment]
Planned for today.

**Comment by m-kuhn** [comment]
Good job!

I also like the preview.

Removing an attribute breaks the `mDefaultExpressionMap` since it's index-based?
What do you think about switching this to name-based instead?
Also what do you think about putting this into QgsFields, I think it would be a better suited place to store it.

**Comment by nyalldawson** [comment]
> What do you think about switching this to name-based instead?

Done!

> Also what do you think about putting this into QgsFields, I think it would be a better suited place to store it.

Great idea! I've *partially* done this, by moving the default expression + alias to QgsFeature. I've had to keep the QMaps around for now (didn't want to modify the dataproviders so that they'd have to track the alias/default values), but it's a step in the right direction...

**Comment by m-kuhn** [comment]
Amazing feature, thanks a lot.
I think we are only just scratching the surface of the possibilities it opens up.
I already wonder when and how we'll integrate this with update (`last_modification_timestamp`) and which variables will be available in all the contexts it could be used in :)

**Comment by nirvn** [comment]
@nyalldawson , here's a cool possibility: using the newly added "collect" aggregation method, automatically input a field value based on a spatial overlap with a given layer's geometries.

If aggregate() function gets support for a way to accent the parent feature from which aggregation is being called, we could even have a land cover polygon layer from which a newly entered {point,string,plygon} has a land_cover field automatically filed based on spatial overlap. Something like: aggregate(layer:="land_cover",aggregate:="concatenate",expression:="cover_type", filter:= intersects($geometry, geometry(@parent_feature))) 

As @m-kuhn said, this opens up so many doors.

**Pull request merged**